src/goto-instrument/havoc_utils.cpp

```diff
@@ -27,47 +29,37 @@ void append_safe_havoc_code_for_expr(
   const auto skip_target =
     skip_program.add(goto_programt::make_skip(source_location));
 
-  // for deref expressions, check for validity of underlying pointer,
-  // and skip havocing if invalid (to avoid spurious pointer deref errors)
-  if(expr.id() == ID_dereference)
-  {
-    const auto condition = not_exprt(w_ok_exprt(
-      to_dereference_expr(expr).pointer(),
-      from_integer(0, unsigned_int_type())));
-    dest.add(goto_programt::make_goto(skip_target, condition, source_location));
-  }
+  // skip havocing only if all pointer derefs in the expression are valid
+  // (to avoid spurious pointer deref errors)
+  dest.add(goto_programt::make_goto(
+    skip_target, not_exprt{validate_all_derefs(expr, ns)}, source_location));
 
   havoc_code_impl();
 
-  // for deref expressions, add the final skip target
-  if(expr.id() == ID_dereference)
-    dest.destructive_append(skip_program);
+  // add the final skip target
+  dest.destructive_append(skip_program);
 }
 
 void append_object_havoc_code_for_expr(
   const source_locationt source_location,
   const exprt &expr,
   goto_programt &dest)
 {
-  append_safe_havoc_code_for_expr(source_location, expr, dest, [&]() {
-    codet havoc(ID_havoc_object);
-    havoc.add_source_location() = source_location;
-    havoc.add_to_operands(expr);
-    dest.add(goto_programt::make_other(havoc, source_location));
-  });
+  codet havoc(ID_havoc_object);
+  havoc.add_source_location() = source_location;
+  havoc.add_to_operands(expr);
+  dest.add(goto_programt::make_other(havoc, source_location));
 }
 
 void append_scalar_havoc_code_for_expr(
   const source_locationt source_location,
   const exprt &expr,
   goto_programt &dest)
 {
-  append_safe_havoc_code_for_expr(source_location, expr, dest, [&]() {
-    side_effect_expr_nondett rhs(expr.type(), source_location);
-    goto_programt::targett t = dest.add(
-      goto_programt::make_assignment(expr, std::move(rhs), source_location));
-    t->code_nonconst().add_source_location() = source_location;
-  });
+  side_effect_expr_nondett rhs(expr.type(), source_location);
```

Comment:
IF this function is specific for scalars, we should add a precondition to make sure `expr` is a scalar.

Reply:
@SaswatPadhi ?